ibm/service/power/resource_ibm_pi_network_security_group_rule.go

```diff
@@ -0,0 +1,611 @@
+									Attr_Maximum: {
+										Computed:    true,
+										Description: "The end of the port range, if applicable, If values are not present then all ports are in the range.",
+										Type:        schema.TypeFloat,
+									},
+									Attr_Minimum: {
+										Computed:    true,
+										Description: "The start of the port range, if applicable. If values are not present then all ports are in the range.",
+										Type:        schema.TypeFloat,
+									},
+								},
+							},
+							Type: schema.TypeList,
+						},
+					},
+				},
+				Type: schema.TypeList,
+			},
+			Attr_UserTags: {
+				Computed:    true,
+				Description: "The user tags associated with this resource.",
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Type:        schema.TypeList,
+			},
+		},
+	}
+}
+
+func resourceIBMPINetworkSecurityGroupRuleCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sess, err := meta.(conns.ClientSession).IBMPISession()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	cloudInstanceID := d.Get(Arg_CloudInstanceID).(string)
+	nsgClient := instance.NewIBMIPINetworkSecurityGroupClient(ctx, sess, cloudInstanceID)
+	nsgID := d.Get(Arg_NetworkSecurityGroupID).(string)
+
+	if v, ok := d.GetOk(Arg_NetworkSecurityGroupRuleID); ok {
+		ruleID := v.(string)
+		_, err := nsgClient.DeleteRule(nsgID, ruleID)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		_, err = isWaitForIBMPINetworkSecurityGroupRuleRemove(ctx, nsgClient, nsgID, ruleID, d.Timeout(schema.TimeoutDelete))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		d.SetId(fmt.Sprintf("%s/%s", cloudInstanceID, nsgID))
+	} else {
+		action := d.Get(Arg_Action).(string)
+		name := d.Get(Arg_Name).(string)
+
+		networkSecurityGroupAddRule := models.NetworkSecurityGroupAddRule{
+			Action: &action,
+			Name:   &name,
+		}
+
+		// Add protocol
+		protocol := d.Get(Arg_Protocol).(map[string]interface{})
+		networkSecurityGroupAddRule.Protocol = networkSecurityGroupRuleMapToProtocol(protocol)
+
+		// Add remote
+		remote := d.Get(Arg_Remote).(map[string]interface{})
+		networkSecurityGroupAddRule.Remote = networkSecurityGroupRuleMapToRemote(remote)
+
+		// Optional fields
+		destinationPort := d.Get(Arg_DestinationPorts).(map[string]interface{})
+		networkSecurityGroupAddRule.DestinationPorts = networkSecurityGroupRuleMapToPort(destinationPort)
+
+		sourcePort := d.Get(Arg_SourcePorts).(map[string]interface{})
+		networkSecurityGroupAddRule.SourcePorts = networkSecurityGroupRuleMapToPort(sourcePort)
+
+		networkSecurityGroup, err := nsgClient.AddRule(nsgID, &networkSecurityGroupAddRule)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		_, err = isWaitForIBMPINetworkSecurityGroupRuleAdd(ctx, nsgClient, nsgID, *networkSecurityGroup.ID, d.Timeout(schema.TimeoutCreate))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		d.SetId(fmt.Sprintf("%s/%s/%s", cloudInstanceID, nsgID, *networkSecurityGroup.ID))
```

Comment:
`nsgID == *networkSecurityGroup.ID` We want the rule id that was just created but we can't since the api does not return it :sweat: